ocaml/xapi/storage_impl.ml

```diff
@@ -183,7 +185,16 @@ module Vdi = struct
 
   let dp_on_vdi dp t = List.mem_assoc dp t.dps
 
-  let get_dp_state dp t =
+  let dp_on_vdi_by_id dp t =
+    List.mem_assoc dp (List.map (fun ((dp, _vm), st) -> (dp, st)) t.dps)
```

Comment:
`List.mem_assoc` uses physical quality rather than structural quality - is this the right choice? `assoc` can also fail if the element is not present; is this guaranteed not to happen? Otherwise `List.assoc_opt` is a better choice.

Reply:
In this case, I was modifying an existing function so that there was a function uisng the new dp structure and the old string

Reply:
`mem_assoc` looks wrong to me - we should discuss this - also if it exists in existing code.

Reply:
`List.mem_assoc` does use structural equality; it is `List.mem_assq` that uses physical.